authn/token_exchange.go

```
@@ -0,0 +1,211 @@
+	if c.cfg.usesSystemCAP && (tokenReq.OrgID == 0 || len(tokenReq.Realms) == 0) {
+		return "", fmt.Errorf("org ID and realms must be specified when using system-wide CAP")
+	}
+
+	key, err := tokenExchangeCacheKey(tokenReq)
+	if err != nil {
+		return "", fmt.Errorf("failed to generate cache key: %v", err)
+	}
+	if res, err := c.cache.Get(ctx, key); err == nil {
+		token := string(res)
+		return token, nil
+	}
+
+	resp, err, _ := c.singlef.Do(key, func() (interface{}, error) {
+		target, err := url.JoinPath(c.cfg.AuthAPIURL, tokenExchangePath)
+		if err != nil {
+			return nil, err
+		}
+
+		req, err := http.NewRequestWithContext(ctx, http.MethodPost, target, strings.NewReader("{}"))
+		if err != nil {
+			return nil, err
+		}
+		req.Header.Set("Authorization", "Bearer "+c.cfg.CAP)
+		req.Header.Set("Content-Type", "application/json")
+		req.Header.Set("Accept", "application/json")
+		if len(tokenReq.Realms) > 0 {
+			realms, err := json.Marshal(tokenReq.Realms)
+			if err != nil {
+				return nil, fmt.Errorf("failed to marshal realms: %v", err)
+			}
+			req.Header.Set("X-Realms", string(realms))
+		}
+		if tokenReq.OrgID != 0 {
+			req.Header.Set("X-Org-ID", strconv.Itoa(int(tokenReq.OrgID)))
+		}
+
+		res, err := c.client.Do(req)
+		if err != nil {
+			return nil, fmt.Errorf("failed to fetch access token: %v", err)
+		}
+
+		defer res.Body.Close()
+
+		response := tokenExchangeResponse{}
+		if err := json.NewDecoder(res.Body).Decode(&response); err != nil {
+			return nil, err
+		}
+
+		if res.StatusCode != http.StatusOK || response.Status != "success" {
+			return nil, fmt.Errorf("status code: %d, error: %s", res.StatusCode, response.Error)
+		}
+
+		return response, nil
+	})
+	if err != nil {
+		return "", fmt.Errorf("failed to fetch access token: %v", err)
+	}
+
+	response, ok := resp.(tokenExchangeResponse)
+	if !ok {
+		return "", fmt.Errorf("unexpected response type")
+	}
+
+	c.cacheValue(ctx, response.Data.Token, key)
```

Did you intend to ignore the error here? I suppose there's a little bit of value in handling the error since it would enforce that the received token is a valid JWT.

---

I was thinking about it, but it felt like it's not worth returning an error if we simply fail to cache the token. I've since refactored the code, so `cacheValue ` would also return an error if the token is not a valid JWT, so I think you're right that it makes sense to return an error instead of swallowing it. I'll do that for now, and we can change it later if needed.